### PR TITLE
Merge 2.12 to 2.13 aug 7 [ci: last-only]

### PR DESCRIPTION
**Pull request by lrytz**
- #6828 is fixed in 2.13 already (PR description)
- #6788 is a backport
- #6754 is NOT a backport, even though it mentions #6670 in the description. #6670 was closed.
- #6865 is a backport
- #6949 updates a backported test

```
➜  scala2 git:(2.13.x) git log upstream/2.12.x --oneline -n 1 | cat
90c94cc2ea Merge pull request #6949 from som-snytt/issue/t10956-fix

➜  scala2 git:(2.13.x) git log HEAD --oneline -n 1 | cat
39c5d37147 Merge pull request #7003 from joshlemer/issue/11050-2.13

➜  scala2 git:(2.13.x) git checkout -b merge-2.12-to-2.13-aug-7
Switched to a new branch 'merge-2.12-to-2.13-aug-7'

➜  scala2 git:(merge-2.12-to-2.13-aug-7) export MB=$(git merge-base HEAD upstream/2.12.x)

➜  scala2 git:(merge-2.12-to-2.13-aug-7) git log --graph --oneline $MB..upstream/2.12.x | cat
*   90c94cc2ea Merge pull request #6949 from som-snytt/issue/t10956-fix
|\
| * e37f3f3afb Avoid truncation of REPL output
|/
*   9a8328e468 Merge pull request #6933 from adriaanm/t11020
|\
| * 1fd64fa0a4 Respect BaseTypeSeq invar (3) symbols are distinct
|/
*   f552875c17 Merge pull request #6919 from retronym/ticket/10858
|\
| * f3cad7e721 Fix regression with curried implicit defs conforming to function types
|/
*   cf4fb702b1 Merge pull request #6882 from jhnaldo/2.12.x
|\
| * 393a2661d6 Fix error in HashMap.HashMapCollision1.merge0
* |   cbfcdb6339 Merge pull request #6865 from som-snytt/issue/10956-backport
|\ \
| * | d9cedb599c [nomerge] Backport test
| * | 6dda47dcec [nomerge] Backport REPL big print fix
| |/
* |   0d4624123a Merge pull request #6696 from rorygraves/mike/2.12.x_rangepos_pr
|\ \
| * | 49bb79f3f2 Optimise rangepos checking use traverser rather than external iteration reuse temporary results where possible
| * | 37d0ca05b9 Unit test for range position validation
|/ /
* |   21c593e818 Merge pull request #6754 from hrhino/topic/rangepos-no-validate_2.12
|\ \
| |/
|/|
| * d736b3bded Move range position validation behind a flag.
* |   d322514e16 Merge pull request #6846 from hrhino/t9529
|\ \
| * | 3e00af4525 Support @Repeatable Java annotations.
* | |   b29d891b8f Merge pull request #6850 from NthPortal/bug#10747/PR
|\ \ \
| * | | 8128fa790b bug#10747 Document thread-safety of Enumeration
|/ / /
* | |   f29c55b424 Merge pull request #6848 from hrhino/faster/getParts
|\ \ \
| |/ /
|/| |
| * | dc0b8b2015 Minor optimizations in implicit search.
|/ /
* |   ec21aad47c Merge pull request #6828 from retronym/ticket/10703
|\ \
| |/
|/|
| * 5af85b5ce6 [nomerge] Several entries with the same key in mutable.HashMap
|/
*   aa9f79f03a Merge pull request #6788 from lrytz/b10487
|\
| * 04cd5c392c Consider delayedInit$body classes as top-level for Java compatibility
|/
* 5622f5d4a2 Merge pull request #6763 from hrhino/t10937
* 24f1438693 Parenthesize infix type constructors before using them as a prefix.


➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge aa9f79f03a
➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge ec21aad47c -s ours
➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge d322514e16
CONFLICT (content): Merge conflict in test/files/run/reify_ann1b.scala
CONFLICT (content): Merge conflict in test/files/run/reify_ann1b.check
CONFLICT (content): Merge conflict in test/files/pos/annotations.scala
CONFLICT (content): Merge conflict in src/reflect/scala/reflect/runtime/JavaUniverseForce.scala
CONFLICT (modify/delete): src/reflect/mima-filters/2.12.0.forwards.excludes deleted in HEAD and modified in d322514e16. Version d322514e16 of src/reflect/mima-filters/2.12.0.forwards.excludes left in tree.
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/typechecker/RefChecks.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/typechecker/Implicits.scala

➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge 0d4624123a
CONFLICT (content): Merge conflict in test/junit/scala/tools/nsc/symtab/SymbolTableForUnitTesting.scala
CONFLICT (content): Merge conflict in src/reflect/scala/reflect/internal/Positions.scala

➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge cbfcdb6339 -s ours

➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge 9a8328e468
CONFLICT (content): Merge conflict in test/junit/scala/collection/immutable/HashMapTest.scala

➜  scala2 git:(merge-2.12-to-2.13-aug-7) git merge 90c94cc2ea -s ours
```

### CODE REVIEW

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala [hunk 1]
```diff
@@ -1177,11 +1177,12 @@ trait Implicits {
      *  bound, the implicits infos which are members of these companion objects.
      */
     private def companionImplicitMap(tp: Type): InfoMap = {
+      val isScala213 = settings.isScala213
 
       /* Populate implicit info map by traversing all parts of type `tp`.
        * Parameters as for `getParts`.
        */
-      def getClassParts(tp: Type)(implicit infoMap: InfoMap, seen: mutable.Set[Type], pending: Set[Symbol]) = tp match {
+      def getClassParts(tp: Type)(implicit infoMap: InfoMap, seen: mutable.HashSet[Type], pending: Set[Symbol]) = tp match {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Implicits.scala:1188]: @hrhino you might want to compare the change here to https://github.com/scala/scala/pull/6848/files#diff-7c03397456d3b987549fd039d6b639c6R1079

File: src/compiler/scala/tools/nsc/typechecker/RefChecks.scala [hunk 2]
```diff
@@ -1391,18 +1391,46 @@ abstract class RefChecks extends Transform {
...
+        annots
+          .map(_.transformArgs(transformTrees))
+          .groupBy(_.symbol)
+          .flatMap((groupRepeatableAnnotations _).tupled)
+          .toList
       }
 
+      // assumes non-empty `anns`
+      def groupRepeatableAnnotations(sym: Symbol, anns: List[AnnotationInfo]): List[AnnotationInfo] =
+        if (!sym.isJavaDefined) anns else anns match {
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:1410]: @hrhino this is slightly different than in 2.12.x (https://github.com/scala/scala/pull/6846/files#diff-e2cfc09e4d7c753232666c6e31bd1926R1424)

Comment (hrhino) [reply src/compiler/scala/tools/nsc/typechecker/RefChecks.scala:1410]: yep, thanks for adapting it!

File: src/compiler/scala/tools/nsc/typechecker/Implicits.scala (hunk 1 above)
Comment (hrhino) [reply src/compiler/scala/tools/nsc/typechecker/Implicits.scala:1188]: 👍

**Review by hrhino — APPROVED** [review]

**Pull request merged**